nomos-cli/src/da/network/swarm.rs

```diff
@@ -129,3 +129,78 @@ where
         }
     }
 }
+
+#[cfg(test)]
+pub mod test {
+    use crate::da::network::swarm::ExecutorSwarm;
+    use crate::test_utils::AllNeighbours;
+    use kzgrs_backend::common::blob::DaBlob;
+    use kzgrs_backend::common::Column;
+    use libp2p::identity::Keypair;
+    use libp2p::PeerId;
+    use nomos_da_network_core::address_book::AddressBook;
+    use nomos_da_network_core::swarm::validator::ValidatorSwarm;
+    use nomos_libp2p::Multiaddr;
+    use std::time::Duration;
+    use tokio::sync::mpsc::unbounded_channel;
+
+    #[tokio::test]
+    async fn test_dispersal_with_swarms() {
+        let k1 = Keypair::generate_ed25519();
+        let k2 = Keypair::generate_ed25519();
+        let executor_peer = PeerId::from_public_key(&k1.public());
+        let validator_peer = PeerId::from_public_key(&k2.public());
+        let neighbours = AllNeighbours {
+            neighbours: [
+                PeerId::from_public_key(&k1.public()),
+                PeerId::from_public_key(&k2.public()),
+            ]
+            .into_iter()
+            .collect(),
+        };
+
+        let addr: Multiaddr = "/ip4/127.0.0.1/udp/5063/quic-v1".parse().unwrap();
+        let addr2 = addr.clone().with_p2p(validator_peer).unwrap();
+        let addr2_book = AddressBook::from_iter(vec![(executor_peer, addr2.clone())]);
+
+        let (dispersal_broadcast_sender, _) = unbounded_channel();
+
+        let mut executor =
+            ExecutorSwarm::new(k1, neighbours.clone(), dispersal_broadcast_sender.clone());
+        let (mut validator, _) = ValidatorSwarm::new(k2, neighbours.clone(), addr2_book);
+
+        tokio::spawn(async move {
+            let validator_swarm = validator.protocol_swarm_mut();
+            validator_swarm.listen_on(addr).unwrap();
+
+            validator.run().await;
+        });
+
+        tokio::time::sleep(Duration::from_secs(1)).await;
+        executor.dial(addr2).unwrap();
```

Comment:
After dialing, executor needs to signal the other end about new stream:
```
let executor_open_stream_sender = executor.open_stream_sender();
executor_open_stream_sender.send(*validator_peer).unwrap();
 ```
 To have `open_stream_sender` you'd need to target the `da-integration-tests` branch.

Reply:
I could not wrap my head around one problem. When I call `executor_disperse_blob_sender.send()`, handle_dispersal_event() is never called from run().loop. What else should I do ?

Reply:
I feel like our Swarm abstractions are only wrappers around libp2p's concepts? And we just forward all good and bad into upper layers.

Reply:
> I feel like our Swarm abstractions are only wrappers around libp2p's concepts? And we just forward all good and bad into upper layers.

Well, they are. This is just a networking layer, it should not do anything else that forward/received messages according to the specs.

Reply:
> I could not wrap my head around one problem. When I call executor_disperse_blob_sender.send(), handle_dispersal_event() is never called from run().loop. What else should I do ?

If you inspect the executor dispersal behaviour, you'll see that before the blob is sent, the `peer_id` for desired subnetwork need to be available (`.../executor/behaviour.rs:289`). For this reason try to apply the changes I suggested and call `open_stream_sender()`.

Reply:
> > I could not wrap my head around one problem. When I call executor_disperse_blob_sender.send(), handle_dispersal_event() is never called from run().loop. What else should I do ?
> 
> If you inspect the executor dispersal behaviour, you'll see that before the blob is sent, the `peer_id` for desired subnetwork need to be available (`.../executor/behaviour.rs:289`). For this reason try to apply the changes I suggested and call `open_stream_sender()`.

Why do I need to call open_stream_sender() ? I know the blob got dispersed, but it looks our event values travel somewhere else ?

Reply:
I can see here, there is even loop similar to what exist in swarms        [ async fn test_sampling_swarm(](https://github.com/logos-co/nomos-node/blob/9f4f139771e96adaad9f365dab7303cace887a0c/nomos-da/network/core/src/protocols/sampling/mod.rs#L96) It is like we always need to `swarm.next().await` and whatever we put in goes out ?

Reply:
> > > > I feel like our Swarm abstractions are only wrappers around libp2p's concepts? And we just forward all good and bad into upper layers.
> > > 
> > > 
> > > Well, they are. This is just a networking layer, it should not do anything else that forward/received messages according to the specs.
> > 
> > 
> > Let me study Swarms in lipp2p more. Both go-waku and nwaku don't use them, that's why I need to have a deeper look.
> 
> You can think of Swarm as a hook or controller for all the underlying behaviours.

I feel uncomfortable we bring channels/hooks into higher levels. Channels are like pipes, no house has them visible unless it is for aestetical reasons :)

Reply:
Let me invest one more day and finish this exercise. Perhaps it will be useful for error case tests which require deep understanding of streams.

Reply:
> > > > > I feel like our Swarm abstractions are only wrappers around libp2p's concepts? And we just forward all good and bad into upper layers.
> > > > 
> > > > 
> > > > Well, they are. This is just a networking layer, it should not do anything else that forward/received messages according to the specs.
> > > 
> > > 
> > > Let me study Swarms in lipp2p more. Both go-waku and nwaku don't use them, that's why I need to have a deeper look.
> > 
> > 
> > You can think of Swarm as a hook or controller for all the underlying behaviours.
> 
> I feel uncomfortable we bring channels/hooks into higher levels. Channels are like pipes, no house has them visible unless it is for aestetical reasons :)

Here is an example how things get confusing:
Once `dispersal_broadcast_sender` became `dispersal_events_sender`. Just one level away we already start forgetting and shifting. Let my un-smartness be a guide here. 

https://github.com/logos-co/nomos-node/blob/9f4f139771e96adaad9f365dab7303cace887a0c/nomos-cli/src/da/network/swarm.rs#L47

https://github.com/logos-co/nomos-node/blob/9f4f139771e96adaad9f365dab7303cace887a0c/nomos-cli/src/da/network/backend.rs#L95

@danielSanchezQ  Could not we have something like `Event Manager` which understands messages, errors? Backend user could just drop data and it will be taken care of ? Any performance penalty for this one more redirect ? WDYT ?

Reply:
> Just one level away we already start forgetting and shifting
what do you mean by this?

Reply:
> > Just one level away we already start forgetting and shifting
> > what do you mean by this?

Channel with the same purpose has different names.

Reply:
> > > Just one level away we already start forgetting and shifting
> > > what do you mean by this?
> 
> Channel with the same purpose has different names.

Well, yeah, we could make it more consistent in naming. But having the channels go away the layer should be good. I may be missing the point, but I do not think having an extra layer (event manager) would solve anything.

We usually have 3 layers:

```
-----------------------------
 Core (behaviours + Swarm)
-----------------------------
Backend (network backend service)
-----------------------------
Service
-----------------------------
```

The most important one is the Service, as the API should not change at all if possible. Imagine we change the core for something else (my most famous exaple: **Pigeons**). We just need to create a new (pigeons) backend while the service is left untouched. That makes pieces to be changed easily, is basically changing the type in the final overwatch app.

What I wanted to say with all this is that:
* Wrapping types is ok for the service and backend, if changing types is just a matter of rewrap (into the stream from whatever).
* Escaping types are not, otherwise is difficult to refactor.